torsiondrive/qm_engine.py

```diff
@@ -327,6 +327,151 @@ def load_native_output(self, filename='output.dat'):
         m.build_topology()
         return m
 
+class EngineGaussian(QMEngine):
+    def __init__(self, input_file=None, work_queue=None, native_opt=False, extra_constraints=None):
+        super().__init__(input_file, work_queue, native_opt, extra_constraints)
+        # Check which version of gaussain we have access to
+        for version in ['g16', 'g09']:
+            try:
+                result = subprocess.run([version, 'test.com'], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+                if 'Unable to open input' in result.stdout.decode('utf-8'):
+                    self.gaussian_version = version
+            except FileNotFoundError:
+                pass
```

Comment:
Thanks for making the effort on the auto-selection of gaussian versions. I have 2 suggestions:
1. Putting this functionality in the initializer of the Engine creates ambiguity. Assuming multiple engine instances may run on different machines, one has `g16` and another has `g09`, they will silently use different versions. Also we would like the creation of engine to be as simple as possible. Therefore I think its better to put this functionality in `launch.py` in `create_engine()` function, then pass the version as an additional argument here, such as
```
class EngineGaussian(QMEngine):
    def __init__(self, input_file=None, work_queue=None, native_opt=False, extra_constraints=None, exe=None):
        super().__init__(input_file, work_queue, native_opt, extra_constraints)
        if exe in ('g16', 'g09'):
            self.gaussian_exe = exe
        else:
            raise ValueError("Only g16 and g09 are supported")
```

Reply:
2. In `create_engine()`, you can use the convenient method `shutil.which(program)` to check if `g16` or `g09` can be found. Such as:
```
if enginename == 'gaussian':
    if shutil.which('g16') is not None:
        gaussian_exe = 'g16'
    elif shutil.which('g09') is not None:
        gaussian_exe = 'g09'
    else:
        raise RuntimeError("Neither of g16 and g09 are found, please check your environment")
    engine = EngineGaussian(inputfile, work_queue, native_opt=native_opt, extra_constraints=extra_constraints, exe=gaussian_exe)
else:
    engine = engine_dict[enginename](inputfile, work_queue, native_opt=native_opt, extra_constraints=extra_constraints)    
```
This implementation introduces a little inconsistency in the engine creation function and I can improve that later by adding the `exe` optional keyword for all engines.